TheJumpCloud/support
# SA-2644: Public Function Parallel Implementation_Phase2

## Issues
* [SA-2644](https://jumpcloud.atlassian.net/browse/SA-2644) - SA-2644: Public Function Parallel Implementation_Phase2

## What does this solve?
- Adjust error handling in Get-JCResults
    - Previously, if errors were encountered in a parallel instance, the error was recorded as a string and placed in an array then returned to the session
    - Now, if an error is encountered, the whole error object is passed to an error array that is then returned to the session
        - This will allow end users to access the error objects to determine what caused the error more accurately, it also looks cleaner
- Adds parallel functionality to the following public functions:
    - Get-JCGroup
    - Get-JCCommand
    - Get-JCCommandTarget
- Added Get-JCResults tests for Groups, Commands and CommandResults

## Is there anything particularly tricky?
- Get-JCCommandTarget returns a custom object, in order for the function to truly run in parallel, the custom object should be created in parallel if the session qualifies for it

## How should this be tested?
- Ensure no errors are thrown and the objects returned are properly formed when running the following commands:
    - Get-JCGroup
        - `Get-JCGroup`
        - `Get-JCGroup -Type "User"`
        - `Get-JCGroup -Type "User" -Name "UserGroupName"`
        - `Get-JCGroup -Type "System" `
        - `Get-JCGroup -Type "System" -Name "SystemGroupName"`
    - Get-JCCommand
        - `Get-JCCommand`
        - `Get-JCCommand -CommandID <commandID>`
        - `Get-JCCommand -name <CommandName>`
        - `Get-JCCommand | Get-JCCommand -ByID` 
    - Get-JCCommandTarget
        - `Get-JCCommandTarget -CommandID <commandID>`
        - `Get-JCCommandTarget -CommandID <commandID> -Groups`
